File: Source/Extensions/Tessa.Extensions.Default.Client/Tiles/AcquaintanceTileExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Acquaintance;
using Tessa.Extensions.Default.Shared.Workflow.KrPermissions;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Extensions.Default.Shared.Workflow.Wf;
using Tessa.Localization;
using Tessa.Platform;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.Scheme;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;
using Tessa.UI.Tiles.Extensions;
using Tessa.Views;
using Tessa.Views.Metadata;
using Tessa.Views.Metadata.Criteria;

namespace Tessa.Extensions.Default.Client.Tiles
{
    public sealed class AcquaintanceTileExtension :
        TileExtension
    {
        #region Fields

        private readonly IKrTypesCache krTypesCache;
        private readonly ICardMetadata cardMetadata;
        private readonly ICardRepository cardRepository;
        private readonly CreateCardModelFuncAsync createCardModelFuncAsync;
        private readonly ICardDialogManager dialogManager;
        private readonly IUIHost uiHost;
        private readonly IKrAcquaintanceManager acquaintanceManager;

        #endregion

        #region Constructors

        public AcquaintanceTileExtension(
            IKrTypesCache krTypesCache,
            ICardMetadata cardMetadata,
            ICardRepository cardRepository,
            CreateCardModelFuncAsync createCardModelFuncAsync,
            ICardDialogManager dialogManager,
            IUIHost uiHost,
            IKrAcquaintanceManager acquaintanceManager)
        {
            this.krTypesCache = krTypesCache;
            this.cardMetadata = cardMetadata;
            this.cardRepository = cardRepository;
            this.createCardModelFuncAsync = createCardModelFuncAsync;
            this.dialogManager = dialogManager;
            this.uiHost = uiHost;
            this.acquaintanceManager = acquaintanceManager;
        }

        #endregion

        #region Private Methods

        private static void EnableOnCardUpdateAndNotTaskCard(object sender, TileEvaluationEventArgs e)
        {
            ICardEditorModel editor = e.CurrentTile.Context.CardEditor;
            ICardModel model;

            e.SetIsEnabledWithCollapsing(
                e.CurrentTile,
                editor != null
                && (model = editor.CardModel) != null
                && model.Card.StoreMode == CardStoreMode.Update
                && model.CardType.ID != DefaultCardTypes.WfTaskCardTypeID);
        }

        private async ValueTask<bool> CanUseResolutionsAsync(ICardModel model, CancellationToken cancellationToken = default)
        {
            KrComponents components = await WfHelper.GetUsedComponentsAsync(this.krTypesCache, model.Card, cancellationToken);
            return components.Has(KrComponents.Resolutions);
        }

        /// <summary>
        /// Показывает окно с выбором списка ролей, которым нужно будет отправить ознакомление.
        /// </summary>
        /// <param name="obj"></param>
        private async void ShowAcquaintanceWindowAsync(object obj)
        {
            ICardEditorModel editor = UIContext.Current.CardEditor;
            if (editor == null)
            {
                return;
            }

            // Сперва сохраняем карточку
            ICardModel mainCardModel;
            if ((mainCardModel = editor.CardModel) != null)
            {
                bool cardIsNew = mainCardModel.Card.StoreMode == CardStoreMode.Insert;

                if (!cardIsNew && (await mainCardModel.HasChangesAsync() || NotEnoughPermissions(editor.CardModel, out KrToken krToken)))
                {
                    await KrTileHelper.OpenMarkedCardAsync(
                        KrPermissionsHelper.CalculateResolutionPermissionsMark,
                        null,       // Не требуем подтверждения действия, если не было изменений
                        () => true, // Автоматом подтверждаем сохранение
                        async () =>
                        {
                            // Нужно заново обратиться к модели в редакторе,
                            // т.к. ссылка mainCardModel уже не указывает на актуальный CardModel
                            if (NotEnoughPermissions(editor.CardModel, out krToken))
                            {
                                TessaDialog.ShowNotEmpty(
                                    ValidationResult.FromText(
                                        this,
                                        KrPermissionsHelper.GetNotEnoughPermissionsErrorMessage(
                                            KrPermissionFlagDescriptors.CreateResolutions),
                                        ValidationResultType.Error));

                                return false;
                            }

                            await DispatcherHelper.InvokeInUIAsync(() =>
                                OpenRolesDialogAsync(
                                    editor,
                                    this.cardMetadata,
                                    this.cardRepository,
                                    this.dialogManager,
                                    this.uiHost,
                                    this.acquaintanceManager,
                                    this.createCardModelFuncAsync));

                            return true;
                        });
                }
                else
                {
                    OpenRolesDialogAsync(
                        editor,
                        this.cardMetadata,
                        this.cardRepository,
                        this.dialogManager,
                        this.uiHost,
                        this.acquaintanceManager,
                        this.createCardModelFuncAsync);
                }
            }
        }

        private static bool NotEnoughPermissions(ICardModel cardModel, out KrToken krToken)
        {
            krToken = KrToken.TryGet(cardModel.Card.Info);
            return krToken != null && !krToken.HasPermission(KrPermissionFlagDescriptors.CreateResolutions);
        }

        private static async void OpenRolesDialogAsync(
            ICardEditorModel editor,
            ICardMetadata cardMetadata,
            ICardRepository cardRepository,
            ICardDialogManager dialogManager,
            IUIHost uiHost,
            IKrAcquaintanceManager acquaintanceManager,
            CreateCardModelFuncAsync createCardModelFuncAsync)
        {
            // Запрашиваем информируемых по умолчанию
            ICardModel model = editor.CardModel;
            if (model == null)
            {
                return;
            }

            var defaultRolesRequest = new CardRequest
            {
                RequestType = DefaultRequestTypes.GetDefaultAcquaintanceRoles,
                CardID = model.Card.ID,
            };

            CardResponse defaultRolesResponse = await cardRepository.RequestAsync(defaultRolesRequest);

            ValidationResult defaultRolesResult = defaultRolesResponse.ValidationResult.Build();
            TessaDialog.ShowNotEmpty(defaultRolesResult);
            if (!defaultRolesResult.IsSuccessful)
            {
                return;
            }

            // Построение диалога
            CardTypeNamedForm dialogForm;
            if (!(await cardMetadata.GetCardTypesAsync()).TryGetValue("Dialogs", out CardType dialogType) ||
                (dialogForm = dialogType.Forms.FirstOrDefault(x => x.Name == "Acquaintance")) == null)
            {
                return;
            }

            CardNewRequest request = new CardNewRequest { CardTypeID = dialogType.ID };
            CardNewResponse response = await cardRepository.NewAsync(request);
            if (!CheckResponseAndSetID(response))
            {
                return;
            }

            ICardModel windowCardModel = await createCardModelFuncAsync(
                response.Card,
                response.SectionRows,
                dialogManager.ShowRowAsync);

            // Получаем список ролей для ознакомления по умолчанию
            List<Tuple<Guid, string>> defaultRoles = AcquaintanceHelper.GetAcquaintanceDefaultRoles(defaultRolesResponse.Info);
            ListStorage<CardRow> rolesRows = windowCardModel.Card.Sections["DialogRoles"].Rows;

            foreach ((Guid roleID, string roleName) in defaultRoles)
            {
                CardRow roleRow = rolesRows.Add();
                roleRow.RowID = Guid.NewGuid();
                roleRow["RoleID"] = roleID;
                roleRow["RoleName"] = roleName;
                roleRow.State = CardRowState.None;
            }

            var windowClosure = new Window[] { null };

            // Показываем диалог
            await uiHost.ShowFormDialogAsync(
                LocalizationManager.Localize(dialogForm.TabCaption),
                dialogForm,
                windowCardModel,
                buttons: new[]
                {
                    new UIButton(
                        "$UI_Common_OK",
                        isDefault: true,
                        buttonActionAsync: async btn =>
                        {
                            var contentElement = windowClosure[0]?.Content as UIElement;
                            if (contentElement != null)
                            {
                                contentElement.IsEnabled = false;
                            }

                            bool closed = false;
                            ValidationResult result;

                            try
                            {
                                if (rolesRows.Count == 0)
                                {
                                    TessaDialog.ShowNotEmpty(ValidationResult.FromText(
                                        "$KrMessages_Acquaintance_RolesRequired", ValidationResultType.Warning));
                                    return;
                                }

                                // Получаем комментарий
                                string comment = windowCardModel.Card.Sections["Dialogs"].Fields.Get<string>("Comment");

                                Card card = model.Card;
                                Guid cardID = card.ID;
                                Guid[] roleIDs = rolesRows.Select(x => x.Get<Guid>("RoleID")).ToArray();

                                using (TessaSplash.Create(TessaSplashMessage.ExecutingCommand))
                                {
                                    result = await acquaintanceManager.SendAsync(cardID, roleIDs, false, comment, addSuccessMessage: true);
                                }

                                if (result.IsSuccessful)
                                {
                                    closed = true;
                                    await btn.CloseAsync();
                                }
                            }
                            finally
                            {
                                if (!closed && contentElement != null)
                                {
                                    contentElement.IsEnabled = true;
                                }
                            }

                            TessaDialog.ShowNotEmpty(result);
                        }),
                    new UIButton("$UI_Common_Cancel", isCancel: true),
                },
                initializeWindowActionAsync: async (window, ct) => windowClosure[0] = window);
        }

        private static bool CheckResponseAndSetID(CardNewResponse response)
        {
            response.Card.ID = Guid.NewGuid();
            return CheckResponse(response);
        }

        private static bool CheckResponse(CardResponseBase response)
        {
            ValidationResult responseResult = response.Validate();
            TessaDialog.ShowNotEmpty(responseResult);
            if (!responseResult.IsSuccessful)
            {
                return false;
            }

            ValidationResult result = response.ValidationResult.Build();
            TessaDialog.ShowNotEmpty(result);
            return result.IsSuccessful;
        }

        /// <summary>
        /// Открывает новую вкладку с представлением AcquaintanceHistory
        /// </summary>
        /// <param name="obj"></param>
        private async void OpenAcquaintanceHistoryViewAsync(object obj)
        {
            ICardEditorModel editor = UIContext.Current.CardEditor;
            ICardModel model;
            if (editor != null
                && !editor.OperationInProgress
                && (model = editor.CardModel) != null)
            {
                Guid cardID = model.Card.ID;

                var parameterMetadata =
                    new ViewParameterMetadata
                    {
                        Alias = "CardIDParam",
                        Caption = LocalizationManager.GetString("Views_Acquaintance_CardID"),
                        Hidden = true,
                        SchemeType = SchemeType.Guid,
                        Multiple = false
                    };

                var parameters =
                    new List<RequestParameter>
                    {
                        new RequestParameterBuilder()
                            .WithMetadata(parameterMetadata)
                            .AddCriteria(new EqualsCriteriaOperator(), cardID.ToString("B"), cardID)
                            .AsRequestParameter()
                    };

                await this.uiHost.ShowViewAsync(
                    "AcquaintanceHistory",
                    LocalizationManager.GetString("Views_AcquaintanceHistory"),
                    parameters);
            }
        }

        #endregion

        #region Base Overrides

        public override Task InitializingGlobal(ITileGlobalExtensionContext context)
        {
            ITileContextSource contextSource = context.Workspace.LeftPanel;

            var tile = new Tile(
                TileNames.AcquaintanceGroup,
                TileHelper.SplitCaption("$KrTiles_AcquaintanceGroup"),
                context.Icons.Get("Thin83"),
                contextSource,
                DelegateCommand.Empty,
                TileGroups.Cards,
                order: 28,
                evaluating: EnableOnCardUpdateAndNotTaskCard,
                tiles: new TileCollection
                {
                    new Tile(
                        TileNames.Acquaintance,
                        TileHelper.SplitCaption("$KrTiles_Acquaintance"),
                        context.Icons.Get("Thin83"),
                        context.Workspace.LeftPanel,
                        new DelegateCommand(this.ShowAcquaintanceWindowAsync),
                        TileGroups.Cards,
                        order: 1,
                        evaluating: EnableOnCardUpdateAndNotTaskCard),

                    new Tile(
                        TileNames.AcquaintanceHistory,
                        TileHelper.SplitCaption("$KrTiles_AcquaintanceHistory"),
                        context.Icons.Get("Thin84"),
                        context.Workspace.LeftPanel,
                        new DelegateCommand(this.OpenAcquaintanceHistoryViewAsync),
                        TileGroups.Cards,
                        order: 2,
                        evaluating: EnableOnCardUpdateAndNotTaskCard)
                });

            tile.SetActionsGrouping();
            context.Workspace.LeftPanel.Tiles.Add(tile);

            return Task.CompletedTask;
        }


        public override async Task InitializingLocal(ITileLocalExtensionContext context)
        {
            ITilePanel panel = context.Workspace.LeftPanel;
            ICardEditorModel editor;
            ICardModel model;
            ITile acquaintanceGroup;

            if ((editor = panel.Context.CardEditor) != null
                && (model = editor.CardModel) != null
                && (acquaintanceGroup = panel.Tiles.TryGet(TileNames.AcquaintanceGroup)) != null
                && (!(await WfHelper.TypeSupportsWorkflowAsync(this.krTypesCache, model.CardType))
                    || !await this.CanUseResolutionsAsync(model)))
            {
                // дочерние плитки тоже скрываются
                acquaintanceGroup.DisableWithCollapsing();
            }
        }

        #endregion
    }
}

File: WebClient SDK/src/default/tiles/acquaintanceTileExtension.ts
import { UIContext, showNotEmpty, createCardModel, UIButton } from 'tessa/ui';
import {
  TileExtension,
  ITileGlobalExtensionContext,
  Tile,
  TileGroups,
  TileEvaluationEventArgs,
  ITileLocalExtensionContext,
  disableWithCollapsing,
  openMarkedCard
} from 'tessa/ui/tiles';
import { showView, showFormDialog } from 'tessa/ui/uiHost';
import { ICardModel, ICardEditorModel } from 'tessa/ui/cards';
import { CardStoreMode, CardRowState, CardRow } from 'tessa/cards';
import { CardRequest, CardService, CardNewRequest } from 'tessa/cards/service';
import { CardTypeFlags } from 'tessa/cards/types';
import { RequestParameterBuilder } from 'tessa/views';
import { ViewParameterMetadata, equalsCriteriaOperator } from 'tessa/views/metadata';
import { hasFlag, TypedField, DotNetType, createTypedField, Guid } from 'tessa/platform';
import { ValidationResult, ValidationResultType } from 'tessa/platform/validation';
import {
  KrTypesCache,
  KrComponents,
  KrToken,
  getKrComponentsByCard,
  CreateResolutions
} from 'tessa/workflow';
import { LocalizationManager } from 'tessa/localization';
import { MetadataStorage } from 'tessa';
import { SchemeType } from 'tessa/scheme';

export class AcquaintanceTileExtension extends TileExtension {
  //#region TileExtension

  public initializingGlobal(context: ITileGlobalExtensionContext) {
    const panel = context.workspace.leftPanel;
    const contextSource = panel.contextSource;

    const tile = new Tile({
      name: 'AcquaintanceGroup',
      caption: '$KrTiles_AcquaintanceGroup',
      icon: 'ta icon-thin-083',
      contextSource,
      group: TileGroups.Cards,
      order: 28,
      evaluating: AcquaintanceTileExtension.enableOnCardUpdateAndNotTaskCard,
      tiles: [
        new Tile({
          name: 'Acquaintance',
          caption: '$KrTiles_Acquaintance',
          icon: 'ta icon-thin-083',
          contextSource,
          command: AcquaintanceTileExtension.showAcquaintanceWindow,
          group: TileGroups.Cards,
          order: 1,
          evaluating: AcquaintanceTileExtension.enableOnCardUpdateAndNotTaskCard
        }),

        new Tile({
          name: 'AcquaintanceHistory',
          caption: '$KrTiles_AcquaintanceHistory',
          icon: 'ta icon-thin-084',
          contextSource,
          command: AcquaintanceTileExtension.openAcquaintanceHistoryView,
          group: TileGroups.Cards,
          order: 2,
          evaluating: AcquaintanceTileExtension.enableOnCardUpdateAndNotTaskCard
        })
      ]
    });

    tile.info['.actionsGrouping'] = true;
    panel.tiles.push(tile);
  }

  public initializingLocal(context: ITileLocalExtensionContext) {
    const panel = context.workspace.leftPanel;
    const editor = panel.context.cardEditor;
    const acquaintanceGroup = panel.tryGetTile('AcquaintanceGroup');
    if (
      !!editor &&
      !!editor.cardModel &&
      !!acquaintanceGroup &&
      (!AcquaintanceTileExtension.typeSupportsWorkflow(editor.cardModel) ||
        !AcquaintanceTileExtension.canUseResolutions(editor.cardModel))
    ) {
      disableWithCollapsing(acquaintanceGroup);
    }
  }

  //#endregion

  //#region evaluatings

  private static enableOnCardUpdateAndNotTaskCard(e: TileEvaluationEventArgs) {
    const editor = e.currentTile.context.cardEditor;
    e.setIsEnabledWithCollapsing(
      e.currentTile,
      !!editor &&
        !!editor.cardModel &&
        editor.cardModel.card.storeMode === CardStoreMode.Update &&
        editor.cardModel.cardType.id !== 'de75a343-8164-472d-a20e-4937819760ac'
    ); // WfTaskCard
  }

  //#endregion

  //#region actions

  private static async showAcquaintanceWindow() {
    const context = UIContext.current;
    const cardEditor = context.cardEditor;

    if (!cardEditor || !cardEditor.cardModel) {
      return;
    }

    const mainCardModel = cardEditor.cardModel;
    const cardIsNew = mainCardModel.card.storeMode === CardStoreMode.Insert;

    if (
      !cardIsNew &&
      ((await mainCardModel.hasChanges()) ||
        AcquaintanceTileExtension.notEnoughPermissions(mainCardModel))
    ) {
      await openMarkedCard(
        'kr_calculate_resolution_permissions',
        null, // Не требуем подтверждения действия, если не было изменений
        async () => true, // Автоматом подтверждаем сохранение
        async () => {
          if (AcquaintanceTileExtension.notEnoughPermissions(cardEditor.cardModel!)) {
            await showNotEmpty(
              ValidationResult.fromText(
                `${LocalizationManager.instance.localize('$KrMessages_NoPermissionsTo')}\
               ${LocalizationManager.instance.localize('$KrPermissions_CreateResolutions')}`,
                ValidationResultType.Error
              )
            );
            return false;
          }

          AcquaintanceTileExtension.openRolesDialog(cardEditor);
          return true;
        }
      );
    } else {
      AcquaintanceTileExtension.openRolesDialog(cardEditor);
    }
  }

  private static openAcquaintanceHistoryView() {
    const context = UIContext.current;
    const cardEditor = context.cardEditor;

    if (cardEditor && cardEditor.cardModel) {
      const cardId = cardEditor.cardModel.card.id;

      const parameterMetadata = new ViewParameterMetadata();
      parameterMetadata.alias = 'CardIDParam';
      parameterMetadata.caption = '$Views_Acquaintance_CardID';
      parameterMetadata.hidden = true;
      parameterMetadata.schemeType = SchemeType.Guid;
      parameterMetadata.multiple = false;

      const parameters = [
        new RequestParameterBuilder()
          .withMetadata(parameterMetadata)
          .addCriteria(equalsCriteriaOperator(), cardId, cardId)
          .asRequestParameter()
      ];

      showView({
        viewAlias: 'AcquaintanceHistory',
        displayValue: '$Views_AcquaintanceHistory',
        parameters
      });
    }
  }

  private static canUseResolutions(model: ICardModel): boolean {
    const usedComponents = getKrComponentsByCard(model.card, KrTypesCache.instance);
    return hasFlag(usedComponents, KrComponents.Resolutions);
  }

  private static typeSupportsWorkflow(model: ICardModel): boolean {
    return (
      hasFlag(model.cardType.flags, CardTypeFlags.AllowTasks) &&
      KrTypesCache.instance.cardTypes.some(x => x.id === model.cardType.id)
    );
  }

  private static notEnoughPermissions(model: ICardModel): boolean {
    const krToken = KrToken.tryGet(model.card.info);
    return !!krToken && !krToken.hasPermission(CreateResolutions);
  }

  private static async openRolesDialog(editor: ICardEditorModel) {
    const model = editor.cardModel;
    if (!model) {
      return;
    }

    // Запрашиваем информируемых по умолчанию
    const defaultRolesRequest = new CardRequest();
    defaultRolesRequest.requestType = '7a2cb692-7a55-4519-b193-0ce2de435523'; // GetDefaultAcquaintanceRoles
    defaultRolesRequest.cardId = model.card.id;

    const defaultRolesResponse = await CardService.instance.request(defaultRolesRequest);
    const defaultRolesResult = defaultRolesResponse.validationResult.build();
    await showNotEmpty(defaultRolesResult);
    if (!defaultRolesResult.isSuccessful) {
      return;
    }

    // Построение диалога
    const dialogsType = MetadataStorage.instance.cardMetadata.getCardTypeByName('Dialogs');
    if (!dialogsType) {
      return;
    }

    const dialogForm = dialogsType.forms.find(x => x.name === 'Acquaintance');
    if (!dialogForm) {
      return;
    }

    const request = new CardNewRequest();
    request.cardTypeId = dialogsType.id;
    const response = await CardService.instance.new(request);
    response.card.id = Guid.newGuid();
    await showNotEmpty(response.validationResult.build());
    if (!response.validationResult.isSuccessful) {
      return;
    }

    const windowCardModel = createCardModel(response.card, response.sectionRows);

    // Получаем список ролей для ознакомления по умолчанию
    const defaultRoles: { id: guid; name: string }[] = [];
    const defaultRolesResponseInfo = defaultRolesResponse.info;
    const defaultRolesDictionary = defaultRolesResponseInfo['.DefaultRoles'];
    if (defaultRolesDictionary) {
      const idsList = defaultRolesDictionary['IDList'] as TypedField<DotNetType.Guid>[];
      const namesList = defaultRolesDictionary['NameList'] as TypedField<DotNetType.String>[];
      if (idsList && namesList && idsList.length === namesList.length) {
        for (let i = 0; i < idsList.length; i++) {
          defaultRoles.push({
            id: idsList[i].$value,
            name: namesList[i].$value
          });
        }
      }
    }

    const rolesRows = windowCardModel.card.sections.get('DialogRoles')!.rows;
    for (const role of defaultRoles) {
      const roleRow = new CardRow();
      roleRow.rowId = Guid.newGuid();
      roleRow.set('RoleID', role.id, DotNetType.Guid);
      roleRow.set('RoleName', role.name, DotNetType.String);
      roleRow.state = CardRowState.None;
      rolesRows.push(roleRow);
    }

    await showFormDialog(dialogForm, windowCardModel, null, [
      new UIButton('$UI_Common_OK', async btn => {
        if (rolesRows.length === 0) {
          await showNotEmpty(
            ValidationResult.fromText(
              '$KrMessages_Acquaintance_RolesRequired',
              ValidationResultType.Warning
            )
          );
          return;
        }

        // Получаем комментарий
        const comment = windowCardModel.card.sections.get('Dialogs')!.fields.get('Comment') || '';

        const card = model.card;
        const cardId = card.id;

        // Запрос на отправку данных для массового ознакомления
        const sendAcquaintanceRequest = new CardRequest();
        sendAcquaintanceRequest.requestType = '87e36c4a-0cb5-4226-8580-c339b4bbf2b7'; // Acquaintance
        sendAcquaintanceRequest.cardId = cardId;

        sendAcquaintanceRequest.info['.Comment'] = createTypedField(comment, DotNetType.String);
        sendAcquaintanceRequest.info['.Roles'] = rolesRows.map(x => x.getField('RoleID')!);
        sendAcquaintanceRequest.info['.ExcludeDeputies'] = createTypedField(
          false,
          DotNetType.Boolean
        );
        sendAcquaintanceRequest.info['.AddSuccessMessage'] = createTypedField(
          true,
          DotNetType.Boolean
        );
        // sendAcquaintanceRequest.info['.PlaceholderAliases']
        // sendAcquaintanceRequest.info['.AdditionalInfo'];

        const sendAcquaintanceResponse = await CardService.instance.request(
          sendAcquaintanceRequest
        );
        await showNotEmpty(sendAcquaintanceResponse.validationResult.build());
        if (!sendAcquaintanceResponse.validationResult.isSuccessful) {
          return;
        }

        btn.close();
      }),
      new UIButton('$UI_Common_Cancel', btn => {
        btn.close();
      })
    ]);
  }

  //#endregion
}
